File: src/models/StudyFile.js
import mongoose from 'mongoose';

const StudyFileSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
    index: true,
  },
  fileName: {
    type: String,
    required: true,
    trim: true,
  },
  fileType: {
    type: String,
    enum: ['pdf', 'image', 'doc', 'spreadsheet', 'other'],
    default: 'other',
  },
  mimeType: {
    type: String,
    required: true,
  },
  fileSize: {
    type: Number,
    required: true,
  },
  fileData: {
    type: String, // base64 encoded — max ~10MB
    required: true,
  },

  // AI-generated metadata
  title: {
    type: String,
    trim: true,
    maxlength: 200,
  },
  summary: {
    type: String,
    default: '',
    maxlength: 3000,
  },

  // Organization
  category: {
    type: String,
    enum: ['system-design', 'dsa', 'web-dev', 'database', 'devops', 'math', 'college', 'notes', 'other'],
    default: 'other',
  },
  tags: [{
    type: String,
    trim: true,
    lowercase: true,
  }],

  isPinned: {
    type: Boolean,
    default: false,
  },
  isArchived: {
    type: Boolean,
    default: false,
  },
}, { timestamps: true });

StudyFileSchema.index({ userId: 1, category: 1, createdAt: -1 });
StudyFileSchema.index({ userId: 1, isPinned: -1, createdAt: -1 });

export default mongoose.models.StudyFile || mongoose.model('StudyFile', StudyFileSchema);

File: backend-spring/src/main/java/com/nextlevel/api/model/StudyFile.java
package com.nextlevel.api.model;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.CompoundIndexes;
import org.springframework.data.mongodb.core.index.CompoundIndex;
import org.springframework.data.mongodb.core.mapping.Document;

import com.fasterxml.jackson.annotation.JsonView;

import lombok.Data;

@Data
@Document(collection = "studyfiles")
@CompoundIndexes({
        @CompoundIndex(def = "{'userId': 1, 'category': 1, 'createdAt': -1}"),
        @CompoundIndex(def = "{'userId': 1, 'isPinned': -1, 'createdAt': -1}")
})
public class StudyFile {
    @Id
    @JsonView(Views.Public.class)
    private String id;
    
    @JsonView(Views.Public.class)
    private String userId;
    
    @JsonView(Views.Public.class)
    private String fileName;
    
    @JsonView(Views.Public.class)
    private String fileType = "other";
    
    @JsonView(Views.Public.class)
    private String mimeType;
    
    @JsonView(Views.Public.class)
    private Long fileSize;
    
    @JsonView(Views.Internal.class)
    private String fileData;
    
    @JsonView(Views.Public.class)
    private String title;
    
    @JsonView(Views.Public.class)
    private String summary = "";
    
    @JsonView(Views.Public.class)
    private String category = "other";
    
    @JsonView(Views.Public.class)
    private List<String> tags = new ArrayList<>();
    
    @JsonView(Views.Public.class)
    private Boolean isPinned = false;
    
    @JsonView(Views.Public.class)
    private Boolean isArchived = false;
    
    @JsonView(Views.Public.class)
    private Instant createdAt;
    
    @JsonView(Views.Public.class)
    private Instant updatedAt;
}
